text/0000-relax-mixin-contract.md

```diff
@@ -0,0 +1,69 @@
+
+This RFC proposed that we update the buildpack specification to create a new relationship between `mypkg`, `run:mypkg`, and `build:mypkg`.
+
+- While `run:mypkg` may imply a different set of changes to the run image compared to `mypkg`, buildpacks requiring `mypkg` should be satisfied with `run:mypkg` at runtime (and vice versa).
+
+- While `build:mypkg` may imply a different set of changes to the build image compared to `mypkg`, buildpacks requiring `mypkg` should be satisfied with `build:mypkg` at build-time (and vice versa).
+
+- A runtime stackpack would still be passed either `run:mypkg` or `mypkg` depending on what the buildpack required, so that the stackpack understands whether or not it can expect the build image to have the corresponding mixin as well.
+
+- A build-time stackpack would still be passed either `build:mypkg` or `mypkg` depending on what the buildpack required, so that the stackpack understands whether or not it can expect the run image to have the corresponding mixin as well.
+
+# Definitions
+[definitions]: #definitions
+
+- [**mixins** and **stage specifiers**](https://github.com/buildpacks/spec/blob/main/platform.md#mixins)
+- [**contractual compatibility** driven by ABI](https://github.com/buildpacks/spec/blob/main/platform.md#compatibility-guarantees)
+
+
+# Motivation
+[motivation]: #motivation
+
+1. Stack authors must currently provide both stage-specific and non-stage-specific versions of mixin names on each image to ensure that buildpacks can require mixins with or without the stage specifier. This is unfriendly UX for stack authors, and difficult to explain.
+
+2. Stack authors are unable to relax the contract between the build and run image (by only using the stage-specifier version of the mixins) without requiring the buildpacks manually list the mixin twice with the different stage specifiers.
+
+3. Under the pre-RFC mixin definition, stackpacks (specified in #87) will attempt to re-install packages when stage-specifier is mismatched between the buildpack and stack images. For instance, a buildpack requesting `mypkg` would result in the stackpack attempting to install `mypkg` on a run image with `run:mypkg`.
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Pack and lifecycle will relax contractual restrictions when initiating builds, creating builders, and exporting images.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This change reduces the power of the mixin model for stack authors. Previously, stack authors could reliably use `run:mypkg` to indicate that a run image strictly contains the runtime version of mypkg (which may provide different functionality than `mypkg`).
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Refactor mixins entirely, e.g., enforce stack<->stack and stack<->buildpack mixin interfaces separately.
```

Comment:
I am inclined to say we should either go this route or remove phase specific prefixes from `io.buildpacks.stack.mixins` entirely (keeping them in buildpack requirements to indicate whether they are required on one or both image).

The way I see it, the distinction between the changsets described by`build:x` and `x` either truly matter (in which case we should be enforcing the difference and refactoring the interface to make it easier to understand the difference) or they don't really matter, in which case we should remove the distinction between the contracts entirely.

If we are in the latter case, while it is true that some build image changes have an implied requirement for some run image changes, I think it think it is sufficient for a buildpack to use mixin `mypkg` to indicate both halves of that contract. The existance of build image mixins shouldn't introduce a requirement for complementary run-image mixins unless at least one buildpack actually depends on the mixin.

Reply:
Is coming up with this new interface worth blocking #87, vs. relaxing the restriction for now?

Reply:
I think for now we should remove `build:` and `run:` entirely from `io.buildpacks.stack.mixins` (which is functionally similar to relaxing the requirement) and then reintroduce it later in a more thoughtful manner if we feel the need for it. I think the phase specific mixin UX is currently confusing to users and relaxing the requirement doesn't go far enough.

Reply:
> I think for now we should remove build: and run: entirely from io.buildpacks.stack.mixins

Not sure I understand. Wouldn't this be a huge breaking change that also forces stack image authors to include many additional, unnecessary packages in run images?

I'm very opposed to losing functionality we already have. This proposal just iterates what we already have into something that isn't more complex. While I totally believe we can find a more elegant solution to the problem, I don't see a disadvantage to making the change proposed.

Reply:
> Wouldn't this be a huge breaking change that also forces stack image authors to include many additional, unnecessary packages in run images?

I am not suggesting that we get rid of `run:pkg` declarations in buildpacks stack requirements, but merely get rid of them in the stack image labels (i.e.`run:pkg` means `pkg` is required but only on the run image).

Reply:
That would break validating compatibility between the build image and run image.

Reply:
Keeping intra-stack contracts for now, as discussed.